### PR TITLE
Cody Lint: polished hackathon version

### PR DESCRIPTION
**Pull request by RXminuS**

### CODE REVIEW

**Comment by RXminuS** [comment]
* **#5137** <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/5137?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @RXminuS and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: agent/src/cli/command-lint.ts [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+- Apply multiple rules to individual files:
+    cody lint --rule '/path/to/pr-pr-rules.codylint.yaml' --rule 'path/to/security.codylint.yaml' --file 'path/to/file1.ts' --file 'path/to/file2.ts'
+
+- Apply rules to a diff:
+    git diff --name-only | cody lint --rule '/path/to/my-rules.codylint.yaml.' --stdin
+`
+        )
+        .option('-i, --file <files...>', 'Code files to lint')
+        .option('-r, --rule <rules...>', 'Codylint files listing rules')
+        .option('-o, --output <output>', 'Output file')
```
Comment (olafurpg) [inline agent/src/cli/command-lint.ts:56]: What format is the output? I'm guessing we will add some kind of `--output-format` later.

File: agent/src/cli/command-lint.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+//TODO: Remove duplication from `command-chat.ts`
+const loginInstruction = 'Sign in with the command: cody auth login --web'
+
+export const lintCommand = () =>
+    new Command('lint')
+        .summary('Apply custom lint rules to your code')
+        .description(
+            `Use Cody to apply custom lint rules to your code.
+
+Examples:
```
Comment (olafurpg) [inline agent/src/cli/command-lint.ts:42]: Very helpful 👍🏻

File: agent/src/cli/command-lint.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+            //TODO: Need a more robust way
+            const stdinPathsPromise = options.stdin
+                ? new Promise<string[]>(resolve => {
+                      const lines: string[] = []
+                      process.stdin.on('data', data => lines.push(...data.toString().split('\n')))
+                      process.stdin.on('end', () => resolve(lines.filter(line => line.trim() !== '')))
+                  })
+                : undefined
+            if (!options.accessToken) {
```
Comment (olafurpg) [inline agent/src/cli/command-lint.ts:86]: Note to self, I want to provide a helper to do this so we don't copy-paste so much code between commands. Should be an easy refactor later, really no need to worry about it here.

File: agent/src/cli/command-root.ts [hunk 4]
```diff
@@ -16,5 +23,12 @@ export const rootCommand = new Command()
     )
     .addCommand(authCommand())
     .addCommand(chatCommand())
+    .addCommand(lintCommand(), {
+        hidden: !(experimental.all || experimental.lint),
```
Comment (olafurpg) [inline agent/src/cli/command-root.ts:27]: Nice 👍🏻

File: vscode/src/jsonrpc/agent-protocol.ts [hunk 5]
```diff
@@ -305,6 +305,10 @@ export type ClientRequests = {
             state: RemoteRepoFetchState
         },
     ]
+    'lint/demo': [
+        { targetFiles: string[]; lintFiles: string[]; model?: string },
```
Comment (olafurpg) [inline vscode/src/jsonrpc/agent-protocol.ts:309]: should probably use "ruleFiles" to mirror `--rule` from the CLI. Also, need to specify these are URIs.

File: agent/src/cli/command-lint.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+    const lintFiles = options.rule
+        .map(file => vscode.Uri.file(path.resolve(process.cwd(), file)))
+        .map(uri => uri.toString())
+    const targetFiles = (options.file ?? [])
+        .concat(options.filesArgs ?? [], options.stdinPaths ?? [])
+        .map(file => vscode.Uri.file(path.resolve(process.cwd(), file)))
+        .map(uri => uri.toString())
+
+    console.log('targetFiles', targetFiles)
```
Comment (olafurpg) [inline agent/src/cli/command-lint.ts:165]: Remove

File: vscode/src/jsonrpc/agent-protocol.ts [hunk 7]
```diff
@@ -305,6 +305,10 @@ export type ClientRequests = {
...
+        { success: boolean; diagnostics: { uri: string; diagnostics: ProtocolDiagnostic[] }[] },
```
Comment (olafurpg) [inline vscode/src/jsonrpc/agent-protocol.ts:310]: Let's extract the nested anonymous struct into an interface type. Will be much cleaner in the generated bindings for Kotlin/Java.

File: vscode/src/configuration.ts [hunk 8]
```diff
@@ -144,6 +144,7 @@ export function getConfiguration(
 
         experimentalSupercompletions: getHiddenSetting('experimental.supercompletions', false),
         experimentalMinionAnthropicKey: getHiddenSetting('experimental.minion.anthropicKey', undefined),
+        experimentalCodyLintEnabled: getHiddenSetting('experimental.lint.enabled', true),
```
Comment (olafurpg) [inline vscode/src/configuration.ts:147]: Disable by default?
```suggestion
        experimentalCodyLintEnabled: getHiddenSetting('experimental.lint.enabled', false),
```

File: vscode/src/lint/options/menu.ts [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+
+        // if (initialInput.length === 0) {
+        //     // If we have no initial input, we want to ensure we don't auto-select anything
+        //     // This helps ensure the input does not feel like a menu.
+        //     lintInput.input.activeItems = []
+        // }
+    })
+}
+
+// const rangeSymbolsInput = createQuickPick({
```
Comment (olafurpg) [inline vscode/src/lint/options/menu.ts:244]: Dead code

File: vscode/src/lint/options/menu.ts [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+//                 ? acceptedItem.range
+//                 : await acceptedItem.range()
+
+//         updateActiveRange(range)
+//         lintInput.render(lintInput.input.value)
+//     },
+// })
+
+// NOTE: this is base64 encoded because of some special character issues. Really this should be loaded from file.
+const ONBOARDING_LINT_TEMPLATE =
```
Comment (olafurpg) [inline vscode/src/lint/options/menu.ts:322]: Can you elaborate what this is?

File: vscode/src/configuration.test.ts [hunk 11]
```diff
@@ -96,6 +96,8 @@ describe('getConfiguration', () => {
                         return false
                     case 'cody.experimental.minion.anthropicKey':
                         return undefined
+                    case 'cody.experimental.pre-r.enabled':
```
Comment (olafurpg) [inline vscode/src/configuration.test.ts:99]: `pre-r`?

**Review by olafurpg — CHANGES_REQUESTED** [review]
Overall LGTM 👍🏻 Requesting changes for mostly basic changes. Let's get this merged behind a flag and get some feedback!

* Would be nice if we can remove `vscode/scripts/lint/tsguide.codylints.yaml`.
* Might be worth adding a feature flag so we can turn this on for the Sourcegraph org without asking people to fiddle with settings.

**Pull request closed without merging**